{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "system('gunzip *')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### fig 3a\n",
    "\n",
    "rm(list = ls())  \n",
    "\n",
    "options(warn=-1)\n",
    "\n",
    "library(Peptides)\n",
    "\n",
    "# a function returning per residue average hydrophobicity for a group of sequences of varying length as well as the number of sequences counted at each position\n",
    "# seqs: a list of sequences\n",
    "# L_max: take the last L_max residues for very long sequences\n",
    "# L_min: only consider sequences at least L_min aa long.\n",
    "# if a sequence is shorter than L_max, fill the beginning with 'x', which will not be used for calculating average hydrophobicity at that position\n",
    "avg_hydrophobicity_per_position <-function(seqs,L_max,L_min,hydrophobicity_scale){\n",
    "  H <- rep(0,L_max) # hydrophobicity at each position\n",
    "  N <- rep(0,L_max) # count of sequence at each position\n",
    "  i=0\n",
    "  for(seq in seqs){\n",
    "    # print progress for every 1000\n",
    "    i=i+1\n",
    "    if (i%%1000==0) {\n",
    "      print(i)\n",
    "    }\n",
    "    \n",
    "    if(nchar(seq) >= L_min){\n",
    "      # extend the N-terminal with x\n",
    "      seq = paste(paste(rep('x',L_max),collapse=''),seq,sep='')\n",
    "      seq = substr(seq,nchar(seq)-L_max+1,nchar(seq))\n",
    "      \n",
    "      hi = hydrophobicity(unlist(strsplit(seq,'')),hydrophobicity_scale) \n",
    "      H = H + hi \n",
    "      N = as.numeric(N) + as.numeric(hi!=0)\n",
    "    }\n",
    "  }\n",
    "  H = H/N\n",
    "  return(list(H,N))\n",
    "}\n",
    "\n",
    "# same as above but look at N-term\n",
    "avg_hydrophobicity_per_position_nterm <-function(seqs,L_max,L_min,hydrophobicity_scale){\n",
    "  H <- rep(0,L_max) # hydrophobicity at each position\n",
    "  N <- rep(0,L_max) # count of sequence at each position\n",
    "  i=0\n",
    "  for(seq in seqs){\n",
    "    # print progress for every 1000\n",
    "    i=i+1\n",
    "    if (i%%1000==0) {\n",
    "      print(i)\n",
    "    }\n",
    "    \n",
    "    if(nchar(seq) >= L_min){\n",
    "      seq = paste(seq,paste(rep('x',L_max),collapse=''),sep='') # changed\n",
    "      seq = substr(seq,1,L_max) # changed\n",
    "      hi = hydrophobicity(unlist(strsplit(seq,'')),hydrophobicity_scale) \n",
    "      H = H + hi \n",
    "      N = as.numeric(N) + as.numeric(hi!=0)\n",
    "    }\n",
    "  }\n",
    "  H = H/N\n",
    "  return(list(H,N))\n",
    "}\n",
    "\n",
    "L_max=100\n",
    "L_min=10\n",
    "hydrophobicity_scale = 'Miyazawa' \n",
    "#hydrophobicity_scale = 'Ponnuswamy'\n",
    "saved_data = paste('genome-scale-hydrophobicity-',hydrophobicity_scale,'-100aa.RData',sep='')\n",
    "  \n",
    "if(! file.exists(saved_data)){\n",
    "    \n",
    "  # for canonical proteome (cds), only consider those >200 aa\n",
    "  # TBD: some gene has mulitple isoforms. currently considers all\n",
    "  # more human.CDS.fa-proteome.fa | grep -v \">\" | sed 's/^M//g'  | sed 's/*$//g' | sort | uniq | awk 'length($0)>200' > cds-200aa.txt\n",
    "  x=read.table('cds-200aa.txt')\n",
    "  cds = avg_hydrophobicity_per_position_nterm(x$V1,L_max,L_min,hydrophobicity_scale)\n",
    "  \n",
    "  # for 5' UTR: take the longest ORF, remove M and * at beginning and end. then remove those found in the canoical proteome\n",
    "  # more human.mRNA.fa-utr5-proteome-longest.fa | grep -v \">\" | sed 's/^M//g'  | sed 's/*$//g' | sort | uniq > utr5.longest.txt\n",
    "  # python grep_between_files.py utr5.longest.txt human.CDS.fa-proteome.fa\n",
    "  # mv utr5.longest.txt-not-matched.txt utr5.longest.txt\n",
    "  x=read.table('utr5.10aa.txt')\n",
    "  utr5 = avg_hydrophobicity_per_position(x$V1,L_max,L_min,hydrophobicity_scale)\n",
    "  \n",
    "  ## UTR3\n",
    "  # more human.mRNA.fa-utr3-proteome-frame0.fa | grep -v \">\" | sed 's/^M//g'  | sed 's/*$//g' | awk 'length($0)>=10' | sort | uniq > utr3.10aa.txt\n",
    "  x=read.table('utr3.10aa.txt')\n",
    "  utr3 = avg_hydrophobicity_per_position(x$V1,L_max,L_min,hydrophobicity_scale)\n",
    "  \n",
    "  # longest.lncrna.peptide.txt\n",
    "  # more human.lncRNA.fa-lncrna-proteome-longest.fa | grep -v \">\" | sed 's/^M//g'  | sed 's/*$//g' | sort | uniq > lncrna-logest.txt\n",
    "  # python grep_between_files.py lncrna-logest.txt human.CDS.fa-proteome.fa\n",
    "  # mv lncrna-logest.txtnot-matched.txt lncrna-longest.txt\n",
    "  x=read.table('lncrna-longest.txt')\n",
    "  lncrna = avg_hydrophobicity_per_position(x$V1,L_max,L_min,hydrophobicity_scale)\n",
    "  \n",
    "  # intron peptide: all 3 frames considered, >=30aa, uniq\n",
    "  # more human.intron.fa-intron-proteome.fa | grep -v \">\" | sed 's/^M//g'  | sed 's/*$//g' | awk 'length($0)>=30' | sort | uniq > intron.30aa.txt\n",
    "  # python grep_between_files.py intron.30aa.txt human.CDS.fa-proteome.fa\n",
    "  # mv intron.30aa.txtnot-matched.txt intron.30aa.txt\n",
    "  x=read.table('intron.30aa.txt')\n",
    "  intron = avg_hydrophobicity_per_position(x$V1,L_max,L_min,hydrophobicity_scale)\n",
    "  save.image(file=saved_data)\n",
    "} else {\n",
    "  load(file=saved_data)\n",
    "}\n",
    "\n",
    "# plotting: need to adjust ylim if using a different scale\n",
    "\n",
    "if (hydrophobicity_scale == 'Miyazawa'){\n",
    "  ylim=c(5.25,5.7)\n",
    "} else if (hydrophobicity_scale == 'Ponnuswamy'){\n",
    "  ylim=c(0.35,0.65)\n",
    "}\n",
    "\n",
    "dev.off()\n",
    "\n",
    "pdf(paste('fig3a-',hydrophobicity_scale,'.pdf',sep=''),width=3,height=4)\n",
    "\n",
    "L=100\n",
    "skip_last=2\n",
    "x=-L:-1\n",
    "y=cds[[1]][(L_max-L+1):L_max]\n",
    "lo <- loess(y[1:(L-skip_last)]~x[1:(L-skip_last)])\n",
    "plot(x,y,ylim=ylim,pch=16,cex=0.2,col='red',bty='l',ylab='\\nAverage hydrophobicity',xlab='Position relative to C-termini\\n')\n",
    "lines(x, predict(lo,x), col='red', lwd=2)\n",
    "\n",
    "y=utr5[[1]][(L_max-L+1):L_max]\n",
    "lo <- loess(y[1:(L-skip_last)]~x[1:(L-skip_last)])\n",
    "points(x,y,col='blue',pch=16,cex=0.2)\n",
    "lines(x, predict(lo,x), col='blue', lwd=2)\n",
    "\n",
    "y=utr3[[1]][(L_max-L+1):L_max]\n",
    "lo <- loess(y[1:(L-skip_last)]~x[1:(L-skip_last)])\n",
    "points(x,y,col='orange',pch=16,cex=0.2)\n",
    "lines(x, predict(lo,x), col='orange', lwd=2)\n",
    "\n",
    "y=lncrna[[1]][(L_max-L+1):L_max]\n",
    "lo <- loess(y[1:(L-skip_last)]~x[1:(L-skip_last)])\n",
    "points(x,y,col='green',pch=16,cex=0.2)\n",
    "lines(x, predict(lo,x), col='green', lwd=2)\n",
    "\n",
    "y=intron[[1]][(L_max-L+1):L_max]\n",
    "lo <- loess(y[1:(L-skip_last)]~x[1:(L-skip_last)])\n",
    "points(x,y,col='magenta',pch=16,cex=0.2)\n",
    "lines(x, predict(lo,x), col='magenta', lwd=2)\n",
    "\n",
    "abline(a=noncoding_h,b=0,lty=2,lwd=2,col='gray')\n",
    "abline(a=coding_h,b=0,lty=2,lwd=2,col='gray')\n",
    "#legend('topleft',\"Random\",text.col='gray',bty='n')\n",
    "dev.off()\n",
    "\n",
    "#legend(0,0,legend=c(\"CDS\",\"5'UTR/uORF\",\"3'UTR\",\"LncRNA\",\"Intron\"),text.col=c('red','blue','orange','green','magenta'),bty='n')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "system('gzip *.txt')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
